### ISSUE TITLE
$_GET method of admin's zen_draw_form does not support action page other than self

### ISSUE THREAD

**Issue by mc12345678** [zencart/zencart#3922]
https://github.com/zencart/zencart/blob/b3ed7bc02a0d4fbad68e7a2f086b79d05d5ee33a/admin/includes/functions/html_output.php#L245-L247

(Feel free to revise the title to something that might sound "better").

Basically, if I am on page X in the admin and setup a form using `zen_draw_form` where the form uses `get` parameters instead of `post` parameters, if the action page is not page X, then only page X gets loaded/reloaded when the associated submit button is pressed.  This is instead of loading the alternate page (say page Y) passing along the data that was to appear on the link address...

This was "discovered" through a problem with [Additional Product Images](https://www.zen-cart.com/downloads.php?do=file&id=1063) with a solution posted here: https://www.zen-cart.com/showthread.php?149805-Additional-Product-Images-addon&p=1372382#post1372382

Obviously another way "around" this is to have the current page "process" the self request and then take additional action which really would likely be that the code of the alternate page would be brought into the code of the one file or everything to be modified to use posting instead of getting data...

Note the "use" and results of this style of coding in the following ZC core admin files and note that a 'correction' to how `admin/includes/functions/html_output.php` operates would mean that the following may need adjustment:
https://github.com/zencart/zencart/blob/b3ed7bc02a0d4fbad68e7a2f086b79d05d5ee33a/admin/attributes_controller.php#L775
https://github.com/zencart/zencart/blob/b3ed7bc02a0d4fbad68e7a2f086b79d05d5ee33a/admin/products_price_manager.php#L291

These locations may "search" just fine; however, they do not move to that associated page.  Of course I have found similar "issues" with other code (non-ZC core code) that is currently present, but I haven't used/tried the features that would have revealed this problem.

I'm not sure that I follow why the referenced html_output code section was added, but...

FWIW, I have also looked into the potential use of `' get'` (adding a space either before or after the word) to prevent the use of the above code as the form's method will be properly populated, but this results in the admin's index page being loaded upon attempting to submit the form.

**Comment by scottcwilson** [comment]
I'm pretty sure I have been stung by this too.  Thanks for writing it up, @mc12345678. 

Possible documentation impact to:
https://docs.zen-cart.com/dev/code/forms/
https://docs.zen-cart.com/dev/plugins/upgrading_to_1.5/#rewriting-addon-admin-pages-to-use-form-posts-instead-of-gets

**Comment by drbyte** [comment]
> Possible documentation impact to:
> https://docs.zen-cart.com/dev/code/forms/
> https://docs.zen-cart.com/dev/plugins/upgrading_to_1.5/#rewriting-addon-admin-pages-to-use-form-posts-instead-of-gets

No, don't touch the docs. It's a bug.

**Comment by mc12345678** [comment]
FWIW, I read the second link specifically in relation to the described issue and for those reading this thread for review/understanding, the referenced plugin's use of `$_GET` is for passing "filter" type information meeting the "*only* when indicating selection criteria, and NEVER when performing destructive actions or database write operations" conditions.

Now looking at the first link, *Never use GET for forms.* <-- Isn't this a bit of a ***strong*** statement? Never? Never ever? What about: Avoid the use... Should try to never... ...for forms that collect user entered data... Something other than just never (with link then describing "*the exception*")?!

**Comment by drbyte** [comment]
@mc12345678 would this work okay without unwanted side-effects?

```php
function zen_draw_form($name, $action, $href_parameters = '', $method = 'post', $form_parameters = '', $usessl = 'deprecated') {
    $url = zen_href_link($action, (!empty($href_parameters) ? $href_parameters : ''));
    $form = '<form name="' . zen_output_string($name) . '" action="' . $url . '" method="' . zen_output_string($method) . '"';
    if (!empty($form_parameters)) {
        $form .= ' ' . $form_parameters;
    }
    $form .= '>';
    if (strtolower($method) == 'post') $form .= '<input type="hidden" name="securityToken" value="' . $_SESSION['securityToken'] . '">';

    // for get-method forms, if no cmd= exists in form action or parameters, assume current page
    if (strtolower($method) == 'get') {
        $query_keys = zen_get_url_get_params_as_array($url);
        $cmd = array_key_exists('cmd', $query_keys) ? $query_keys['cmd'] : (isset($_GET['cmd']) ? $_GET['cmd'] : 'home');
        $form .= '<input type="hidden" name="cmd" value="' . $cmd . '">';
    }
    return $form;
}

function zen_get_url_get_params_as_array($url)
{
    $query_segments = explode('&', parse_url($url, PHP_URL_QUERY));
    $query_array = [];
    foreach ($query_segments as $segment) {
        $parts = explode('=', $segment);
        $query_array[$parts[0]] = (isset($parts[1]) ? $parts[1] : '');
    }
    return $query_array;
}
```

**Comment by mc12345678** [comment]
All testing that I have done appears to not introduce unwanted side-effects. Haven't seen much that would use the admin side notifier within `zen_href_link`, but that area might be one that might "cause" some sort of issue as the new parser function evaluates the text of the link provided rather than some "internal" or alternate value. Basically, I'm saying that if an admin observer reworks the url to ***not*** include all of the parameters (`cmd` at least specifically), then there would still be a need to modify/ensure that `$_GET['cmd']` were set "appropriately".

That said, the above works to remove the need for forcing `$_GET['cmd']` to be set to the value of `$action` before drawing the form. Further, by calling `zen_href_link` as part of the link generation, it is possible within there to go ahead and set `$_GET['cmd']` as necessary as well.

Now am curious about the accuracy of the this comment:
> // for get-method forms, if no cmd= exists in form action or parameters, assume current page

how is it that `'home'` is considered to be the current page?

I mean, what specifically *forces* `cmd=` to be the current page? This is/was somewhat the problem that was being seen in one case where a store owner entering: `admin/filename.php` was not redirected to `admin/index.php?cmd=filename` as part of loading that page. Therefore, when the `get` form was generated, it `cmd` was not a parameter on the uri… Seems that if there was to be such a redirect, it didn't happen...

**Comment by drbyte** [comment]
> one case where a store owner entering: `admin/filename.php` was not redirected to `admin/index.php?cmd=filename` as part of loading that page. Therefore, when the `get` form was generated, it `cmd` was not a parameter on the uri… Seems that if there was to be such a redirect, it didn't happen...

With GET forms all uri parameters on the action= tag appear to be dropped upon form submit, replaced by all the input fields. (I haven't found an HTTP RFC to confirm this observation though.)
That is essentially what all this GET-parsing code is all about accommodating.

**Comment by mc12345678** [comment]
> > one case where a store owner entering: `admin/filename.php` was not redirected to `admin/index.php?cmd=filename` as part of loading that page. Therefore, when the `get` form was generated, it `cmd` was not a parameter on the uri… Seems that if there was to be such a redirect, it didn't happen...
> 
> With GET forms all uri parameters on the action= tag appear to be dropped upon form submit, replaced by all the input fields. (I haven't found an HTTP RFC to confirm this observation though.)
> That is essentially what all this GET-parsing code is all about accommodating.

The "closest" that I've found is [RFC 1866](https://www.ietf.org/rfc/rfc1866.txt), section 8.2.2. The only "issue" I have with how it is written is with the phrase `starts with the action URI and appends a '?'`. Reason I have "issue" with it is the interpretation of what is meant by the `action URI` where `action="thisfile.php?param1=x&param2=y"`. It is not ***blatantly*** clear that the `action URI` is actually just everything up to the first parameter identification such as `thisfile.php`.  I understand that if the entire text were used and the `?` were appended then there would be two `?`s which would be incorrect...

As to the "why" all of the "GET-parsing code", *I* get it and appreciate the proposed solution that provides better support of what appears to be standard operation for a `get` style form.

**Comment by neekfenwick** [comment]
Just to chime in, I feel it is very misguided to dictate that a form using method GET should only ever target the page it's on.  That stance hasn't been well defended in any argument I've seen presented.

**Comment by mc12345678** [comment]
> Just to chime in, I feel it is very misguided to dictate that a form using method GET should only ever target the page it's on. That stance hasn't been well defended in any argument I've seen presented.

That argument, to my knowledge, has not been made in this particular issue. It was made elsewhere though in response to this same line of inquiry that went unincorporated for nearly a year and remains so. 

Perhaps I should apologize for chiming in on your reported issue. Maybe your concerns would have been taken better if I hadn't identified its existence.

**Comment by neekfenwick** [comment]
> That argument, to my knowledge, has not been made in this particular issue.
Being pedantic, I didn't say it had.  My implication was that it's implied by the current behaviour of the code, that you cannot ask `zen_draw_form()` to generate a GET form that targets an admin page other than the current one.  The point was made, as you say, on the referenced issue #5454 by @scottcwilson that this is by design, and it's that stance that I disagree with.

**Comment by scottcwilson** [comment]
Fixed by #5568.
